res/controllers/common-hid-packet-parser.js

```diff
@@ -393,17 +393,16 @@ class HIDPacket {
             console.error(`HIDPacket.pack - Parsing packed value: invalid pack format ${field.pack}`);
             return;
         }
-        const bytes = this.packSizes[field.pack];
-        const signed = this.signedPackFormats.includes(field.pack);
         if (field.type === "bitvector") {
             const bitVector = /** @type {HIDBitVector} */ (field.value);
-            if (bytes > 1) {
+            if (this.packSizes[field.pack] > 1) {
                 console.error("HIDPacket.pack - Packing multibyte bit vectors not yet supported");
                 return;
             }
-            for (const bit_id in bitVector.bits) {
-                const bit = bitVector.bits[bit_id];
-                data[field.offset] = data[field.offset] | bit.value;
+            const bitVectorKeyArr = Object.keys(bitVector.bits);
+            let bitVectorIdx = bitVectorKeyArr.length;
+            while (bitVectorIdx--) {
+                data[field.offset] |= bitVector.bits[bitVectorKeyArr[bitVectorIdx]].value;
```

Comment:
I don't see how this is a performance improvement.

Reply:
I tried several loop implementations with QSEngine and this was the fastest I found. The improvements were not much, but stable reproduceable. The execution time, of some real COs of my Z2 mapping, dropped some 100us.
The explanation is, that iterators (obj->next) for complex datatypes are slow, and the iteration is of cause executed many times in a loop. Object.keys() is an optimized JS operation executed only once. The loop itself uses only a basic datatype (integer), with basic iterator.
The last and, least important, aspect is that the condition in the while loop is a comparisition against 0. Which is faster than comparing against Arr.length. The drawback is the reversed order of the loop execution - but doesn't matter here.

Reply:
Mhmm, I know these tricks from languages which compile into machine code, but I'm surprised they work even in a jitted language as JS. Assuming both compiles Also, have you done experiments with `for (key in Object.keys())`? I really can't imagine that the manual iteration and indexing is any faster than iterating over the array or even the object. 

Still, I don't think the performance improvement is really worth the additional code verbosity. If that could be eliminated by hiding it in a forEach functor, I would be ok with it, but you'd need to test whether the Jit would inline it so we still get the performance benefits.